File: h5gl/new/src/pan3d/display/particle/Display3DParticle.ts
module Pan3d {
    export class Display3DParticle extends Display3D {
        public visible: boolean;
        public timeline: TimeLine;
        protected _time: number;
        private _beginTime: number;
        public shader: Shader3D;
        public data: ParticleData;

        public bindMatrix: Matrix3D;
        public bindVecter3d: Vector3D;
        public bindScale: Vector3D;


        public invertBindMatrix: Matrix3D;
        public groupMatrix: Matrix3D;

        protected _rotationMatrix: Matrix3D;
        public modelMatrix: Matrix3D;
      

        public isInGroup: boolean = false;
        public groupPos: Vector3D;
        public groupScale: Vector3D;
        public groupRotation: Vector3D;

        constructor(value: Scene3D) {
            super(value);
            this.visible = true;
            this._rotationMatrix = new Matrix3D();
            this.modelMatrix = new Matrix3D();
        }
        public onCreated(): void {

        }
        public setBind($pos: Vector3D, $rotation: Matrix3D, $scale: Vector3D, $invertRotation: Matrix3D, $groupMatrix: Matrix3D): void {
            this.bindVecter3d = $pos;
            this.bindMatrix = $rotation;
            this.bindScale = $scale;
            this.invertBindMatrix = $invertRotation;
            this.groupMatrix = $groupMatrix;
        }

        public setViewCamModeMatr3d():void
        {
     
            var ctx: Context3D = this.scene3D.context3D;
            ctx.setVcMatrix4fv(this.data.materialParam.shader, "viewMatrix", this.scene3D.camera3D.viewMatrix.m);
            ctx.setVcMatrix4fv(this.data.materialParam.shader, "camMatrix", this.scene3D.camera3D.camMatrix3D.m);
            ctx.setVcMatrix4fv(this.data.materialParam.shader, "modeMatrix", this.modelMatrix.m);
        }

        public getMulBindList(): Array<Vector3D> {
            return null;
        }

        public updateMatrix(): void {

            if (!this.bindMatrix) {
                return;
            }
            this.modelMatrix.identity();
            if (!this.groupMatrix.isIdentity) {
                this.posMatrix.append(this.groupMatrix);
            }
            this.modelMatrix.append(this.posMatrix);
            this.modelMatrix.append(this.bindMatrix);

            this.modelMatrix.appendTranslation(this.bindVecter3d.x, this.bindVecter3d.y, this.bindVecter3d.z);

        }
        //特效配置等级显示  是否能显示
        private get cantUseEffectsLev(): boolean {

            return false;
        }
        public updateTime(t: number): void {
            if (this.cantUseEffectsLev) {
                return;
            }
            this._time = t - this._beginTime;
            this._time += this.data._delayedTime; //加上延时 
            this.timeline.updateTime(t);
            this.visible = this.timeline.visible;
            this.posMatrix.identity();
            this.posMatrix.prependScale(this.scaleX * 0.1 * this.bindScale.x * this.data.overAllScale,
                this.scaleY * 0.1 * this.bindScale.y * this.data.overAllScale,
                this.scaleZ * 0.1 * this.bindScale.z * this.data.overAllScale);
 
            this.timeline.updateMatrix(this.posMatrix, this);
        }

        public reset(): void {
            this.timeline.reset();
            this.updateTime(0);
        }

   
      
        public update(): void {
            {
                if (this.visible && this.data != null && this.scene3D != null) {
                    if (this.data.materialParam != null && this.data.materialParam.shader != null) {
                        this.shader = this.data.materialParam.shader;
                        var ctx: Context3D = this.scene3D.context3D;
                        ctx.setProgram(this.shader.program);
                        ctx.setBlendParticleFactors(this.data._alphaMode);
                        ctx.cullFaceBack(this.data.materialParam.material.backCull);
                        this.updateMatrix();
                        this.setMaterialVc();
                        this.setMaterialTexture();
                        this.setVc();
                        this.setVa();
                        this.resetVa();

                    }
                }
            }
        }
        public setVc(): void {

        }
        public setVa(): void {

        }

        public resetVa(): void {

        }

        public setMaterialVc(): void {
            if (!this.data.materialParam) {
                return;
            }
            var dynamicConstList: Array<DynamicConstItem> = this.data.materialParam.dynamicConstList;
            var t: number = this._time % (Scene3D.frameTime * this.data._life);
            for (var i: number = 0; i < dynamicConstList.length; i++) {
                dynamicConstList[i].update(t);
            }
            if (this.data.materialParam.material.fcNum <= 0) {
                return;
            }
            t = t * this.data.materialParam.material.timeSpeed;
            this.data.materialParam.material.update(t);
            var ctx:Context3D=this.scene3D.context3D;
            ctx.setVc4fv(this.data.materialParam.shader, "fc", this.data.materialParam.material.fcData);


        }
        public setMaterialTexture(): void {
            if (!this.data.materialParam) {
                return;
            }
            var ctx:Context3D=this.scene3D.context3D;
            var texVec: Array<TexItem> = this.data.materialParam.material.texList;
            for (var i: number = 0; i < texVec.length; i++) {
                if (texVec[i].isDynamic) {
                    continue;
                }
                ctx.setRenderTexture(this.data.materialParam.shader, texVec[i].name, texVec[i].texture, texVec[i].id, true);
            }
            var texDynamicVec: Array<DynamicTexItem> = this.data.materialParam.dynamicTexList;
            for (var i: number = 0; i < texDynamicVec.length; i++) {
                ctx.setRenderTexture(this.data.materialParam.shader, texDynamicVec[i].target.name, texDynamicVec[i].texture, texDynamicVec[i].target.id, true);
            }
        }

      
        public setAllByteInfo($byte: Pan3dByteArray, version: number = 0): void {
            this.creatData();
            this.data.version = version;
            this.data.setAllByteInfo($byte);
            this.timeline = new TimeLine();
            this.timeline.setAllDataInfo(this.data.timelineData);
            this._beginTime = this.timeline.beginTime;
        }

        public creatData(): void {
            this.data = new ParticleData(this.scene3D);
        }

        public setTimeLine($tl: TimeLine): void {
            this.timeline = $tl;
            this._beginTime = $tl.beginTime;
        }

    }
}

File: testproject/testdemo/app/src/main/java/com/pan3d/display/particle/Display3DParticle.java
package com.pan3d.display.particle;

import com.pan3d.base.Camera3D;
import com.pan3d.base.Scene_data;
import com.pan3d.core.Context3D;
import com.pan3d.display.Display3D;
import com.pan3d.display.particle.ctrl.TimeLine;
import com.pan3d.material.DynamicConstItem;
import com.pan3d.material.DynamicTexItem;
import com.pan3d.material.TexItem;
import com.pan3d.program.Shader3D;
import com.pan3d.scene.Scene3D;
import com.pan3d.vo.Float32Array;
import com.pan3d.vo.Matrix3D;
import com.pan3d.vo.Vector3D;

import java.util.List;


public class Display3DParticle extends Display3D {

    public boolean visible;
    public TimeLine timeline;
    public Shader3D shader3D;
    public Matrix3D modeMatrix;
    protected float _time;
    private float _beginTime;
    public ParticleData data;
    public Matrix3D bindMatrix;
    public Vector3D bindVecter3d;
    public Vector3D bindScale;
    public Matrix3D invertBindMatrix;
    public Matrix3D groupMatrix;

    public Matrix3D modelMatrix;
    public boolean isInGroup ;
    public Vector3D groupPos;
    public Vector3D groupScale;
    public Vector3D groupRotation;
    public Matrix3D rotationMatrix3D;

    private String  TAG="Display3DParticle";


//    private TextureRes testTextureOne;
//    private TextureRes testTextureTwo;
    public  Display3DParticle(Scene3D val  ){

        super(val);

        this.modeMatrix=new Matrix3D();
        this.rotationMatrix3D=new Matrix3D();
        this._time=0;

//        this.testTextureOne =    this.scene3D.textureManager.createTexture(ColorTransition.getImageDataByRandom(new Vector3D( 1,0,0,1)));
//        this.testTextureTwo =    this.scene3D.textureManager.createTexture(ColorTransition.getImageDataByRandom(new Vector3D( 0,0,1,1)));
    }
    public void setBind(Vector3D $pos, Matrix3D $rotation, Vector3D  $scale, Matrix3D $invertRotation, Matrix3D $groupMatrix)
    {
        this.bindVecter3d = $pos;
        this.bindMatrix = $rotation;
        this.bindScale = $scale;
        this.invertBindMatrix = $invertRotation;
        this.groupMatrix = $groupMatrix;
    }
    public void setTimeLine(TimeLine $tl) {
        this.timeline = $tl;
        this._beginTime = $tl.beginTime;
    }
    public void onCreated()
    {

    }
    public void updateTime(float t)
    {
        this._time = t - this._beginTime;
        this._time += this.data._delayedTime; //加上延时
        this.timeline.updateTime(t);
        this.visible = this.timeline.visible;
        this.posMatrix3d.identity();
        float tempScale=0.1f;
        this.posMatrix3d.prependScale(
                this.scaleX * tempScale * this.bindScale.x * this.data.overAllScale,
                this.scaleY * tempScale * this.bindScale.y * this.data.overAllScale,
                this.scaleZ * tempScale * this.bindScale.z * this.data.overAllScale
        );

        this.timeline.updateMatrix(this.posMatrix3d, this);
    }
    public void update()
    {
        if (this.visible&& this.data!=null&& scene3D !=null){
            if( this.data.materialParam!=null &&this.data.materialParam.shader3D!=null){
                this.shader3D=this.data.materialParam.shader3D;
                Context3D ctx= scene3D.context3D;
                ctx.setProgame(this.shader3D.program);
                ctx.setBlendParticleFactors(this.data._alphaMode);
                ctx.cullFaceBack(this.data.materialParam.material.backCull);

                ctx.setBlendParticleFactors(1);
                ctx.disableCullFace();



                this.updateMatrix();
                this.setMaterialVc();
                this.setMaterialTexture();
                this.setVc();
                this.setVa();
                this.resetVa();

            }
        }
    }
    /*
 设置基础透视，镜头，模型矩阵
 */
    public void  setViewCamModeMatr3d()
    {
        Camera3D cam3d=  scene3D.camera3D;
        Context3D ctx= scene3D.context3D;
        ctx.setVcMatrix4fv(this.shader3D,Shader3D.viewMatrix,cam3d.viewMatrix.m);
        ctx.setVcMatrix4fv(this.shader3D,Shader3D.camMatrix,cam3d.camMatrix3D.m);
        ctx.setVcMatrix4fv(this.shader3D,Shader3D.modeMatrix,this.modeMatrix.m);
    }
    public void  setMaterialTexture()
    {
        Context3D ctx= scene3D.context3D;
        List<TexItem> texVec  = this.data.materialParam.material.texList;
        for (int i   = 0; i < texVec.size(); i++) {
            TexItem texItem=texVec.get(i);
            if (texItem.isDynamic) {
                continue;
            }
            ctx.setRenderTexture(this.shader3D,texItem.name,texItem.textureRes.textTureInt,texItem.get_id());
        }
        List<DynamicTexItem>  texDynamicVec  = this.data.materialParam.dynamicTexList;
        for (int i   = 0; i < texDynamicVec.size(); i++) {
            DynamicTexItem dynamicTexItem=  texDynamicVec.get(i);
            TexItem texItem=dynamicTexItem.target;
            if(dynamicTexItem.hasTextureRes()){
                ctx.setRenderTexture(this.shader3D,texItem.name,dynamicTexItem.getTexture(),texItem.get_id());



            }


        }
    }
    public void  reset()
    {
        this.timeline.reset();
        this.updateTime(0);


    }
    public void  setVa()
    {

    }
    public void  setVc()
    {

    }
    public void  resetVa()
    {

    }
    public void  updateMatrix()
    {
        if (this.bindMatrix==null){
            return;
        }
        this.modeMatrix.identity();
        this.modeMatrix.append(this.posMatrix3d);
        this.modeMatrix.append(this.bindMatrix);
        this.rotationMatrix3D.identity();
        this.rotationMatrix3D.appendRotation(rotationX,Vector3D.X_AXIS);
        this.rotationMatrix3D.appendRotation(rotationY,Vector3D.Y_AXIS);
        this.rotationMatrix3D.appendRotation(rotationZ,Vector3D.Z_AXIS);
        this.modeMatrix.appendTranslation(this.bindVecter3d.x,this.bindVecter3d.y,this.bindVecter3d.z);
    }
    public void  setMaterialVc()
    {
        if (this.data.materialParam==null) {
            return;
        }
        List<DynamicConstItem> dynamicConstList= this.data.materialParam.dynamicConstList;

        float t= Math.floorMod ((long) this._time , (long) (Scene_data.frameTime * this.data._life));
        for (int i = 0; i < dynamicConstList.size(); i++) {
            dynamicConstList.get(i).update(t);
        }
        if(this.data.materialParam.material.fcNum <= 0){
            return;
        }
        t = t * this.data.materialParam.material.timeSpeed;
        this.data.materialParam.material.update(t);
        Float32Array fcData= this.data.materialParam.material.fcData;
        Context3D ctx= scene3D.context3D;
        int fcNum=this.data.materialParam.material.fcNum;
//        Log.d(TAG, fcData.get(0)+" "+fcData.get(1)+" "+fcData.get(2)+" "+fcData.get(3));
        ctx.setVc4fv(this.shader3D,"fc",fcNum, fcData.verBuff);

    }


}
